src/Demo/main.c

```diff
@@ -0,0 +1,182 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/// \brief Example program for UART interface
+
+#include "Uart.h"
+#include "UartRegisters.h"
+#include "ByteFifo.h"
+#include "SystemConfig.h"
+#include "Startup_Leon3.h"
+#include <rtems.h>
+
+#define REG_BIT(reg, n) (((reg) >> n) & 1U)
+#define REG_BYTE(reg, n) (((reg) >> (n * 4)) & 0x0FU)
+#define TIMEOUT 4000000
+#define FIFO_SIZE 256
+
+static Uart uart0;
+
+static Uart_Config config;
+
+static int err;
+
+static uint8_t length;
+
+BYTE_FIFO_CREATE(rxByteFifo, FIFO_SIZE);
+
+BYTE_FIFO_CREATE(txByteFifo, FIFO_SIZE);
+
+static inline char
+itoc(uint8_t c)
+{
+    switch(c) {
+        case 10:
+            return 'A';
+        case 11:
+            return 'B';
+        case 12:
+            return 'C';
+        case 13:
+            return 'D';
+        case 14:
+            return 'E';
+        case 15:
+            return 'F';
+        default:
+            return '0' + c;
+    }
+}
+
+static void
+debugAddr(const uint32_t addr)
+{
+    char buff[10] = "0x00000000";
+    for(unsigned char n = 0; n < 8; n++) {
+        uint8_t byte = 0;
+        byte += REG_BYTE(addr, n);
+        buff[9 - n] = itoc(byte);
+    }
+    for(unsigned char n = 0; n < 10; n++) {
+        Uart_write(&uart0, buff[n], TIMEOUT, &err);
+    }
+}
+
+static void
+debugReg(const uint32_t reg)
+{
+    for(char n = 0; n < 32; n++) {
+        uint8_t byte = 0;
+        byte += REG_BIT(reg, n);
+        Uart_write(&uart0, '0' + byte, TIMEOUT, &err);
+    }
+}
+
+static void
+write(const char* text)
+{
+    int n = 0;
+    while(*(text + n) != 0) {
+        Uart_write(&uart0, *(text + n), TIMEOUT, &err);
+        n++;
+        if(n > 32) {
+            return;
+        }
+    }
+}
+
+static void
+testCallback()
+{
+    ++length;
+}
+
+static void
+testSync(Uart* uart)
+{
+    Uart_startup(uart);
+    Uart_getConfig(uart, &config);
+    uart->rxHandler.targetLength = 5;
+    uart->rxHandler.lengthCallback = testCallback;
+    write("uart status: \n");
+    debugAddr(uart->reg->status);
+    write("\n");
+    debugReg(uart->reg->status);
+    write("\n\ncounting characters (x5)\n");
+    write("type q to end:\n_\r");
+    char buf = '\0';
+    do {
+        Uart_write(uart, itoc(length), TIMEOUT, &err);
+        Uart_write(uart, '\r', TIMEOUT, &err);
+        Uart_read(uart, (uint8_t*)&buf, TIMEOUT, &err);
+    } while(buf != 'q' && err == 0);
+    if(err != 0) {
+        Uart_write(uart, '!', TIMEOUT, &err);
+        Uart_write(uart, err, TIMEOUT, &err);
+    }
+    Uart_write(uart, '\n', TIMEOUT, &err);
+    Uart_shutdown(uart);
+}
+
+static void
+testAsync(Uart* uart)
+{
+    ByteFifo_clear(&txByteFifo);
+    ByteFifo_clear(&rxByteFifo);
+    Uart_startup(uart);
+    Uart_getConfig(uart, &config);
```

Comment:
getConfig before setting it? What is done with the returned data?

Reply:
removed (changed method of code testing)